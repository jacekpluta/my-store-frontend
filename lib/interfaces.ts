export interface CurrentUser {
  user: {
    id: string;
    name: string;
    email: string;
    permissions: Array<string>;
    cart: [
      {
        item: IItem;
        quantity: number;
        size: number;
      }
    ];
  };
}

export interface ICartItem {
  id: number;
  quantity: number;
  size: number;
  item: {
    id: number;
    price: number;
    user: null;
    image: string;
    title: string;
    description: string;
    largeImage: string;
  };
  user: {
    id: string;
    name: string;
    email: string;
    permissions: string[];
  };
}

export interface IUser {
  currentUser: {
    user: {
      id: string;
      name: string;
      email: string;
      permissions: Array<string>;
      cart: [
        {
          item: IItem;
          quantity: number;
          size: number;
        }
      ];
    };
  };
}

export interface IItem {
  id: string;
  price: number;
  user: null;
  image: string;
  title: string;
  description: string;
  largeImage: string;
  brand: string;
  category: string;
  gender: string;
}

export interface ItemsProps {
  page: number;
  filters: any[]; //array of string and objects(with numbers)
}

export interface ItemsState {
  filterOptions: {
    skip: number;
    first: number;
    orderBy: string;
    gender_in?: string[] | undefined;
    category_in?: string[] | undefined;
    brand_in?: string[] | undefined;
    price_gte?: number | undefined;
    price_lte?: number | undefined;
  };
  showPickSize: Boolean;
}

export enum Order {
  createdAtDESC = "createdAt_DESC",
  titleDESC = "title_DESC",
  titleASC = "title_ASC",
  priceASC = "price_ASC",
  priceDESC = "price_DESC",
}

export interface NavBarProps {
  itemsCount: number;
  navOpen: boolean;
  main: boolean;
  toggleBar: boolean;
  path: string;
  currentUser: CurrentUser;
  matchedPermissions: string[];
}
