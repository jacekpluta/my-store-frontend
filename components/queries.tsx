import gql from "graphql-tag";

export const CURRENT_USER_QUERY = gql`
  query {
    user {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          price
          image
          title
          description
        }
      }
    }
  }
`;
