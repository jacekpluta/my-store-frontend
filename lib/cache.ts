import { InMemoryCache } from "@apollo/client";
import {
  isCartOpen,
  filters,
  clearFilters,
  addToCartItem,
  isNavOpen,
  cartLocal,
} from "./vars";

export const cache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        cartOpen: {
          read() {
            return isCartOpen();
          },
        },
        navOpen: {
          read() {
            return isNavOpen();
          },
        },
        filters: {
          read() {
            return filters();
          },
        },
        clearFilters: {
          read() {
            return clearFilters();
          },
        },
        addToCartItem: {
          read() {
            return addToCartItem();
          },
        },
        cartLocal: {
          read() {
            return cartLocal();
          },
        },
      },
    },
  },
});
