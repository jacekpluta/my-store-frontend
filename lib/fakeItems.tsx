export const loadingItems = [
  {
    id: 1,
    price: -1,
    image: "loading",
    title: "loading",
    description: "loading",
    largeImage: "loading",
    brand: "loading",
    category: "loading",
    gender: "loading",
  },
  {
    id: 2,
    price: -1,
    image: "loading2",
    title: "loading2",
    description: "loading2",
    largeImage: "loading2",
    brand: "loading2",
    category: "loading2",
    gender: "loading2",
  },
  {
    id: 3,
    price: -1,
    image: "loading3",
    title: "loading3",
    description: "loading3",
    largeImage: "loading3",
    brand: "loading3",
    category: "loading3",
    gender: "loading3",
  },
  {
    id: 4,
    price: -1,
    image: "loading4",
    title: "loading4",
    description: "loading4",
    largeImage: "loading4",
    brand: "loading4",
    category: "loading4",
    gender: "loading4",
  },
  {
    id: 5,
    price: -1,
    image: "loading5",
    title: "loading5",
    description: "loading5",
    largeImage: "loading5",
    brand: "loading5",
    category: "loading5",
    gender: "loading5",
  },
];
