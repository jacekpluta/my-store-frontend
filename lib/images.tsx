export const images = [
  "./images/shoes1.jpg",
  "./images/shoes2.jpg",
  "./images/shoes3.jpg",
];

export const imageMain = "./images/shoes5.jpg";
export const emptyCart = "./images/emptyCart.png";
export const emptyOrders = "./images/emptyOrders.png";

export const infoImages = [
  "./images/info1.jpg",
  "./images/info2.png",
  "./images/info3.jpg",
];

export const logoImages = [
  "./images/logos/logo1.png",
  "./images/logos/logo2.png",
  "./images/logos/logo3.png",
  "./images/logos/logo4.png",
  "./images/logos/logo5.png",
  "./images/logos/logo8.png",
  "./images/logos/logo9.png",
  "./images/logos/logo10.png",
];
