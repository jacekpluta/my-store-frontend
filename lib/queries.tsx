import gql from "graphql-tag";
import { perPage } from "../config";

export const CURRENT_USER_QUERY = gql`
  query {
    user {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          price
          image
          title
          description
        }
      }
    }
  }
`;

export const IS_CART_OPEN_QUERY = gql`
  query IsCartOpen {
    cartOpen @client
  }
`;

export const FILTERS_QUERY = gql`
  query Filters {
    filters @client
  }
`;

export const CLEAR_FILTERS_QUERY = gql`
  query ClearFilters {
    clearFilters @client
  }
`;

export const ADD_TO_CART_ITEM_QUERY = gql`
  query AddToCartItem {
    addToCartItem @client
  }
`;

export const ALL_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}, $orderBy: ItemOrderByInput, $price_gte: Int, $price_lte: Int,$brand_in: [Brand!], $gender_in: [Gender!], $category_in: [Category!]) {
    items(first: $first, skip: $skip, orderBy: $orderBy, where: { price_gte: $price_gte,price_lte: $price_lte,  brand_in: $brand_in, gender_in: $gender_in, category_in: $category_in}) {
      id
      title
      price
      description
      image
      largeImage
      gender
      brand
      category
    }
  }
`;

export const FEATURED_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = 4) {
    items(first: $first, skip: $skip, orderBy: description_ASC) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

export const DISCOUNT_ITEMS_QUERY = gql`
  query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = 4) {
    items(first: $first, skip: $skip, orderBy: id_ASC) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

export const FILTERED_ITEMS_QUERY = gql`
  query FILTERED_ITEMS_QUERY($skip: Int = 0, $first: Int = ${perPage}, $orderBy: ItemOrderByInput, $price: Int) {
    items(first: $first, skip: $skip, orderBy: $orderBy, where: { price: $price }) {
      id
      title
      price
      description
      image
      largeImage
    }
  }
`;

export const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!, $quantity: Int!) {
    addToCart(id: $id, quantity: $quantity) {
      id
    }
  }
`;
