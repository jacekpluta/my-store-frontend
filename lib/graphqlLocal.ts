import gql from "graphql-tag";
import { ApolloCache } from "apollo-cache";
import { Resolvers } from "apollo-client";
import { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION } from "../components/cart";

export const typeDefs = gql`
  extend type Query {
    cartOpen: Boolean!
  }
`;

type ResolverFn = (
  parent: any,
  args: any,
  { cache }: { cache: ApolloCache<any> }
) => any;

interface ResolverMap {
  [field: string]: ResolverFn;
}

interface AppResolvers extends Resolvers {
  // We will update this with our app's resolvers later
}

export const resolvers = {
  Mutation: {
    toggleCart: (
      _,
      //   { cartOpen }: { cartOpen: boolean },
      variables,
      { cache }: { cache: any }
    ): string[] => {
      const { cartOpen } = cache.readQuery({ query: LOCAL_STATE_QUERY });
      const data = {
        data: { cartOpen: !cartOpen },
      };
      cache.writeData(data);
      return data;
    },
  },
};
