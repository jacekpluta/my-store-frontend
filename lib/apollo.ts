import { IncomingMessage, ServerResponse } from "http";
import { useMemo } from "react";

import { endpoint, productionBackendEndpoint } from "../config";

// import { InMemoryCache, NormalizedCacheObject } from "apollo-cache-inmemory";
// import { ApolloClient } from "@apollo/client";
import { typeDefs, resolvers } from "./graphqlLocal";

import { ApolloClient, InMemoryCache } from "@apollo/client";

let apolloClient: ApolloClient<any> | undefined;

export type ResolverContext = {
  req?: IncomingMessage;
  res?: ServerResponse;
};

function createIsomorphLink(context: ResolverContext = {}) {
  if (typeof window === "undefined") {
    const { SchemaLink } = require("apollo-link-schema");
    const { schema } = require("./schema");
    return new SchemaLink({ schema, context });
  } else {
    const { HttpLink } = require("apollo-link-http");
    return new HttpLink({
      uri:
        process.env.NODE_ENV === "development"
          ? endpoint
          : productionBackendEndpoint,
      credentials: "include",
    });
  }
}
const cache = new InMemoryCache();

function createApolloClient(context?: ResolverContext): any {
  return new ApolloClient({
    uri:
      process.env.NODE_ENV === "development"
        ? endpoint
        : productionBackendEndpoint,
    credentials: "include",
    cache: cache,
    resolvers,
    typeDefs,
  });
}

// cache.writeData({
//   data: {
//     cartOpen: false,
//   },
// });

export function initializeApollo(
  initialState: any = null,
  // Pages with Next.js data fetching methods, like `getStaticProps`, can send
  // a custom context which will be used by `SchemaLink` to server render pages
  context?: ResolverContext
) {
  const _apolloClient = apolloClient ?? createApolloClient(context);

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // get hydrated here
  if (initialState) {
    _apolloClient.cache.restore(initialState);
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === "undefined") return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: any) {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
