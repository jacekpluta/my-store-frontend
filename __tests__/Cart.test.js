import Cart, { LOCAL_STATE_QUERY } from "../components/cart/cart";
import { mount } from "enzyme";
import { MockedProvider } from "@apollo/react-testing";
import { CURRENT_USER_QUERY } from "../lib/queries";

import { ApolloConsumer } from "@apollo/client";

const fakeItem = {
  __typename: "Item",
  id: "test123",
  price: 5000,
  user: null,
  image: "dog-small.jpg",
  title: "dogs are best",
  description: "dogs",
  largeImage: "dog.jpg",
};

const fakeCartItem = {
  __typename: "CartItem",
  id: "omg123",
  quantity: 3,
  item: fakeItem,
  user: {
    __typename: "User",
    id: "4234",
    name: "test",
    email: "test@o2.pl",
    permissions: ["ADMIN"],
  },
};

const fakeUser = {
  __typename: "User",
  id: "4234",
  name: "test",
  email: "test@o2.pl",
  permissions: ["ADMIN"],
  cart: [fakeCartItem],
};

const cartMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { user: fakeUser } },
  },
  {
    request: { query: LOCAL_STATE_QUERY },
    result: { data: { cartOpen: true } },
  },
];

it("renders component and maches snap", async () => {
  const wrapper = mount(
    <MockedProvider mocks={cartMocks}>
      <Cart />
    </MockedProvider>
  );

  await new Promise((resolve) => setTimeout(resolve, 0));
  wrapper.update();

  const header = wrapper.find("header");

  expect(wrapper.find("CartItem")).toHaveLength(1);
  expect(header).toMatchSnapshot();
});
