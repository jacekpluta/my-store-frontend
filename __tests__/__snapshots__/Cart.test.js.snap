// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component and maches snap 1`] = `
<header>
  <CloseButton
    onClick={[Function]}
    title="close"
  >
    <button
      className="CloseButton-e02u0t-0 gwSVMv"
      onClick={[Function]}
      title="close"
    >
      Ã—
    </button>
  </CloseButton>
  test
   - Cart
  <p>
    You have 
    1
     item 
     
    in your cart
  </p>
</header>
`;
